flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolBatchSlotRequestTest.java

```diff
@@ -0,0 +1,341 @@
+
+			failAllocation(slotPool, directMainThreadExecutor, allocationIdFuture.get());
+
+			assertThat(slotFuture.isDone(), is(false));
+		}
+	}
+
+	/**
+	 * Tests that a batch slot request won't fail if its resource manager request fails.
+	 */
+	@Test
+	public void testPendingBatchSlotRequestDoesNotFailIfRMRequestFails() throws Exception {
+		final TestingResourceManagerGateway testingResourceManagerGateway = new TestingResourceManagerGateway();
+		testingResourceManagerGateway.setRequestSlotFuture(FutureUtils.completedExceptionally(new FlinkException("Failed request")));
+
+		final ComponentMainThreadExecutor directMainThreadExecutor = ComponentMainThreadExecutorServiceAdapter.forMainThread();
+
+		final Time batchSlotTimeout = Time.milliseconds(1000L);
+		try (final SlotPoolImpl slotPool = new SlotPoolBuilder()
+			.setComponentMainThreadExecutor(directMainThreadExecutor)
+			.setBatchSlotTimeout(batchSlotTimeout)
+			.setResourceManagerGateway(testingResourceManagerGateway)
+			.build()) {
+
+			final CompletableFuture<PhysicalSlot> slotFuture = requestNewAllocatedBatchSlot(slotPool, directMainThreadExecutor, resourceProfile);
+
+			assertThat(slotFuture.isDone(), is(false));
+		}
+	}
+
+	/**
+	 * Tests that a pending batch slot request times out after the last fulfilling slot gets
+	 * released.
+	 */
+	@Test
+	public void testPendingBatchSlotRequestTimeoutAfterSlotRelease() throws Exception {
+		final ComponentMainThreadExecutor directMainThreadExecutor = ComponentMainThreadExecutorServiceAdapter.forMainThread();
+		final ManualClock clock = new ManualClock();
+		final Time batchSlotTimeout = Time.milliseconds(1000L);
+
+		try (final SlotPoolImpl slotPool = new SlotPoolBuilder()
+				.setComponentMainThreadExecutor(directMainThreadExecutor)
+				.setClock(clock)
+				.setBatchSlotTimeout(batchSlotTimeout)
+				.build()) {
+			final ResourceID taskManagerResourceId = offerSlots(slotPool, directMainThreadExecutor, Collections.singletonList(resourceProfile));
+			final CompletableFuture<PhysicalSlot> firstSlotFuture = requestNewAllocatedBatchSlot(slotPool, directMainThreadExecutor, resourceProfile);
+			final CompletableFuture<PhysicalSlot> secondSlotFuture = requestNewAllocatedBatchSlot(slotPool, directMainThreadExecutor, ResourceProfile.UNKNOWN);
+			final CompletableFuture<PhysicalSlot> thirdSlotFuture = requestNewAllocatedBatchSlot(slotPool, directMainThreadExecutor, smallerResourceProfile);
+
+			final List<CompletableFuture<PhysicalSlot>> slotFutures = Arrays.asList(firstSlotFuture, secondSlotFuture, thirdSlotFuture);
+
+			// initial batch slot timeout check
+			advanceTimeAndTriggerCheckBatchSlotTimeout(slotPool, clock, batchSlotTimeout);
+
+			assertThat(CompletableFuture.anyOf(slotFutures.toArray(COMPLETABLE_FUTURES_EMPTY_ARRAY)).isDone(), is(false));
+
+			releaseTaskManager(slotPool, directMainThreadExecutor, taskManagerResourceId);
+
+			clock.advanceTime(1L, TimeUnit.MILLISECONDS);
```

Comment:
This line can probably be removed from the test.

Reply:
Good point, will remove.